mqtt/mqtt-edgehub/src/command.rs

```diff
@@ -58,81 +78,48 @@ impl CommandHandler {
             broker_handle,
             client,
-        }
+            termination_handle,
+            termination_receiver,
+        })
     }
 
     pub fn shutdown_handle(&self) -> Result<ShutdownHandle, ShutdownError> {
-        self.client
-            .shutdown_handle()
-            .map_or(Err(ShutdownError::ClientDoesNotExist), |shutdown_handle| {
-                Ok(ShutdownHandle(shutdown_handle))
-            })
+        Ok(ShutdownHandle(
+            self.termination_handle.clone(),
+            self.client.shutdown_handle()?,
+        ))
     }
 
-    pub async fn run(mut self) -> Result<(), CommandHandlerError> {
+    pub async fn run(mut self) {
         debug!("starting command handler");
-        let subscribe_topics = &[DISCONNECT_TOPIC.to_string()];
-
-        self.subscribe(subscribe_topics).await?;
-
-        while let Some(event) = self
-            .client
-            .try_next()
-            .await
-            .map_err(CommandHandlerError::PollClientFailure)?
-        {
-            if let Err(e) = self.handle_event(event).await {
-                warn!(message = "error processing command handler event", error = %e);
-            }
-        }
-
-        debug!("command handler disconnected");
-
-        Ok(())
-    }
-
-    async fn subscribe(&mut self, topics: &[String]) -> Result<(), CommandHandlerError> {
-        debug!("command handler subscribing to disconnect topic");
-        let subscriptions = topics.iter().map(|topic| proto::SubscribeTo {
-            topic_filter: topic.to_string(),
-            qos: proto::QoS::AtLeastOnce,
-        });
-
-        for subscription in subscriptions {
-            self.client
-                .subscribe(subscription)
-                .map_err(CommandHandlerError::SubscribeFailure)?;
-        }
 
-        let mut subacks: HashSet<_> = topics.iter().map(Clone::clone).collect();
-
-        while let Some(event) = self
-            .client
-            .try_next()
-            .await
-            .map_err(CommandHandlerError::PollClientFailure)?
-        {
-            if let Event::SubscriptionUpdates(subscriptions) = event {
-                for subscription in subscriptions {
-                    if let SubscriptionUpdateEvent::Subscribe(sub) = subscription {
-                        subacks.remove(&sub.topic_filter);
+        loop {
+            match self.client.try_next().await {
+                Ok(Some(event)) => {
+                    if let Err(e) = self.handle_event(event).await {
+                        error!(message = "error processing command handler event", error = %e);
                     }
                 }
+                Ok(None) => {
+                    error!("command handler client disconnected");
 
-                if subacks.is_empty() {
-                    debug!("command handler successfully subscribed to disconnect topic");
-                    return Ok(());
+                    if let Ok(()) = self.termination_receiver.try_recv() {
```

Comment:
I'm not sure if this will work. I believe that client never returns `None`. What you need to do is to use `futures::select!` macro to select b/w `self.termination_receiver.try_recv()` and `self.client.try_next()`. 

Take a look at recent fix that I've made as an example - https://github.com/Azure/iotedge/blob/0cffc25ac206d23c96d594c11cb68062f58a75e7/mqtt/mqtt-broker-tests-util/src/client.rs#L251-L279

Reply:
The client returns none because the client is shutdown when the command handler shutdown() method is called, right? Also, the shutdown works in the test so I think the client has to be returning None.

Reply:
Yeah, I think you're right, it will return `None` in case of shutdown. But then what's the point of `command_handler_shutdown` channel? If client is being shut down we can just return out of the func and CommandHandler is done.

Reply:
Vadim and I discussed offline and he pointed that client is non-recoverable after it returns None, which is only possible via client.shutdown. Therefore we can revert this change to command handler shutdown.